### DISCUSSION TITLE
What role do Dependencies play in the Manifest file?

### DISCUSSION THREAD

**Discussion by ShaneCourtrille** [OrchardCMS/OrchardCore#10437] [Q&A] [answered]
If we look at OrchardCore.Email as an example, the Manifest file has OrchardCore.Resources as a dependency but that project isn't listed in Dependencies/Projects in Visual Studio.  

So why is it listed and when we create our own Modules what criteria do we use for putting something in as a Dependency?

**Comment by hishamco** [comment]
Let me check, may be it's added accidently

**Reply by hishamco** [reply]
There's a reference here https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Modules/OrchardCore.Email/OrchardCore.Email.csproj#L23

**Reply by ShaneCourtrille** [reply]
@hishamco So why aren't other references listed as Dependencies?

**Reply by hishamco** [reply]
If you look here

https://github.com/OrchardCMS/OrchardCore/blob/7032e2e1007d499a67371f33442e0436dee6999b/src/OrchardCore.Modules/OrchardCore.Email/OrchardCore.Email.csproj#L18-L24

There's no modules have been references except `OrchardCore.Resources`, but I think there's no need for it. I created a PR #10439

**Comment by hishamco** [comment, marked as answer]
> when we create our own Modules what criteria do we use for putting something in as a Dependency?

You need to list all the modules that your module depends on

**Reply by ShaneCourtrille** [reply]
Okay that makes sense.  DisplayManagement isn't a module which is why it's not in the list.  Thanks

**Reply by hishamco** [reply]
To make it clear module MUST NOT reference other module in VS, but may depend on other existing modules / features

**Reply by ShaneCourtrille** [reply]
So how is the Dependency used?  How do we know what to put in our own modules Dependency list?

**Reply by Skrypt** [reply]
The right answer is that if you have modules that have a manifest, you can enable them using the Manifest dependencies. So, for example, the Lucene module depends on the ContentTypes module. If you enable the ContentTypes module it won't enable the Lucene feature but if you enable the Lucene feature it will enable the ContentTypes module.

**Reply by ShaneCourtrille** [reply]
@Skrypt Are they optional though? If I have a Module that is always enabled then do I need the Dependencies?

**Reply by Skrypt** [reply]
They are optional

**Reply by hishamco** [reply]
So the enabled module will enabled all the modules that depends on

**Reply by Skrypt** [reply]
Yeah, maybe we should make a distinction between "Feature" and module here. The dependencies in a Manifest.cs file is about enabling dependent "features". So, a module can have multiple "Features" too ... I let you make yourself a picture in your mind.

**Reply by hishamco** [reply]
Ya features, seems we need to be clear about the termenologies

**Comment by deanmarcussen** [comment]
To be clear feature dependencies serve many purposes.

Taking a dependency in your manifest will (as mentioned cause the dependant feature to be enabled when enabling your feature. Equally when disabling the core feature, it will cause your feature to be disabled, as your feature is dependant.

Additionally they serve as an ordering strategy for many things, including but not limited to
- Default startup ordering (can also be overridden with `Startup.Order` properties
- Migration ordering, i.e. your feature's migrations will run after the dependencies
- View / Alternate ordering and overrides

Whether ordering is important, and requires taking a dependency will largely depend on the functionality your module requires.

While taking a nuget dependency in your `csproj` will give you access to the C# references, it will have little to no impact on how your feature behaves with other features.
